### DISCUSSION TITLE
Is it possible to use this queue with any moveable types?

### DISCUSSION THREAD

**Discussion by ptsouchlos** [erez-strauss/lockfree_mpmc_queue#5] [Q&A] [answered]
While going through the code and the associated presentation, it seems that this queue only supports trivial data types. What I didn't fully understand is why this limitation is placed on the queue. Would supporting any moveable type break the requirements that were set out for this queue implementation?

**Comment by erez-strauss** [comment, marked as answer]
In order to provide non-blocking atomic operation on the queue entries, each entry must be 16 bytes or less, as the hardware atomic operations which are non-blocking are limited to 16 bytes, ( DWCAS, see below). The Queue requires 32 bits for internal entry state, which leaves the transferable data type to 12 bytes.
 
To use the queue with movable objects of 12 bytes or less, you can see the queue implementation that transfers std::unique_ptr<> in the header file: [pointer_mpmc_queue.h](https://github.com/erez-strauss/lockfree_mpmc_queue/blob/master/pointer_mpmc_queue.h) - it uses the reset() and release() methods to keep a single owner to the target object.

In case you need to transfer bigger objects, you might want to use the std::unique_ptr<BigObject> as the transferable units.
Note: don't use the std::unique_ptr with deleter functions, as the the size of the pointer might change to 16 bytes.

You might also use shared_ptr which fits into 8 bytes, ( not std::shared_ptr<> which is 16 bytes) for example boost intrusive shared  pointer


References:

- https://en.wikipedia.org/wiki/Double_compare-and-swap
- https://timur.audio/dwcas-in-c#:~:text=DWCAS%20(Double%2Dwidth%20compare%2D,on%20a%2064%2Dbit%20CPU.

**Reply by ptsouchlos** [reply]
I see now, thank you for the thorough answer!
